### PR TITLE
Show callback failures to platform admins (#NOT-425)

### PR DESCRIPTION
**Pull request by jonalport**
image

image

### CODE REVIEW

File: admin/app/notify_client/service_api_client.py [hunk 1]
```diff
@@ -494,5 +494,15 @@ def update_service_data_retention(self, service_id, data_retention_id, days_of_r
     def get_service_data_retention(self, service_id):
         return self.get("/service/{}/data-retention".format(service_id))
 
+    def get_failing_callback_stats(self):
```
Comment (dagoof) [inline admin/app/notify_client/service_api_client.py:497]: thoughts on putting this into a new API client file that refers to the domain of our new model thats being exposed -- `notify_client/callback_failures_client.py` ?

File: api/app/dao/callback_failures_dao.py [hunk 2]
```diff
@@ -28,6 +29,33 @@ def dao_get_callback_failures_by_service_id(service_id):
     return CallbackFailure.query.join(CallbackFailure.notification).filter_by(service_id=service_id)
 
 
+def dao_get_callback_failures_for_all_services(page=1):
+    return CallbackFailure.query.join(
+        Notification, CallbackFailure.notification_id == Notification.id
```
Comment (dagoof) [inline api/app/dao/callback_failures_dao.py:34]: is this join needed / used anywhere? technically this column refers to `NotificationHistory` and not `Notification` however it does work if needed.

File: api/app/service/rest.py [hunk 3]
```diff
@@ -741,6 +746,36 @@ def get_organisation_for_service(service_id):
     return jsonify(organisation.serialize() if organisation else {}), 200
 
 
+@service_blueprint.route('/failing-callback-stats', methods=['GET'])
+def get_callback_failing_stats_for_all_services():
+    stats = dao_get_callback_failing_stats_for_all_services()
+    data = {
+        'totalFailureCount': stats[0],
```
Comment (dagoof) [inline api/app/service/rest.py:753]: you should be able to refer to the label of the column from callback_failures_dao here:

* `stats.total_failure_count` and
* `stats.failed_notification_count`

we use snake_case for API responses

File: admin/app/templates/views/platform-admin/callback_failure_summary.html [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  {% call(item, row_number) list_table(
+    summary_rows,
+    caption="Services with failing callbacks",
+    caption_visible=False,
+    empty_message='Nothing to see',
+    field_headings=['Service', 'Fails'],
+    field_headings_visible=True
+  ) %}
+
+    {{ link_field(item[0], url_for('main.service_dashboard', service_id=item[1])) }}
```
Comment (dagoof) [inline admin/app/templates/views/platform-admin/callback_failure_summary.html:25]: are we able to serialise out these records in a way that lets us have named fields to reference?

item[0] -> callback_failure.service_name
item[1] -> callback_failure.service_id
item[2] -> callback_failure.total_failure_count

**Review by dagoof — CHANGES_REQUESTED** [review]

File: admin/app/notify_client/service_api_client.py (hunk 1 above)
Comment (jonalport) [reply admin/app/notify_client/service_api_client.py]: Sounds good to me!

File: api/app/dao/callback_failures_dao.py (hunk 2 above)
Comment (jonalport) [reply api/app/dao/callback_failures_dao.py]: Yep not needed

File: api/app/service/rest.py (hunk 3 above)
Comment (jonalport) [reply api/app/service/rest.py]: Oh cool, updated

File: admin/app/templates/views/platform-admin/callback_failure_summary.html (hunk 4 above)
Comment (jonalport) [reply admin/app/templates/views/platform-admin/callback_failure_summary.html]: Yep, updated

**Review by dagoof — APPROVED** [review]

**Pull request merged**